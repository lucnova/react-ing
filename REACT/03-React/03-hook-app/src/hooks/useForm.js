/**
 *  USE FORM HOOK
 *      Con este hook puedo manejar campos de un form.
 *      El estado inicial es un objeto donde cada propiedad es un valor de un input
 *
 *      Regresa el estado actual con una funcion para manejar el cambio de un input.
 */

import { useState } from 'react';

export const useForm = (initialState = {}) => {
	const [formValues, setFormValues] = useState(initialState);

	const handleInputChange = ({ target }) => {
		setFormValues({
			...formValues, // Mantener las posibles propiedades anteriores
			[target.name]: target.value, // Poner la nueva propiedad en su valor
		});
	};

	const resetFormValues = () => {
		setFormValues(initialState);
	};

	return [formValues, handleInputChange, resetFormValues]; //Regresa el estado actual con una funcion para manejar el cambio de un input.
};
