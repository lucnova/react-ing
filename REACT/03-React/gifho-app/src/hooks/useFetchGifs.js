/**
 * EL CUSTOM HOOK HACE EL LLAMADO DE getGifs
 * 
 *  Hace el trabajo pesado, llama al helper para obtener los datos, los agarra y los reenvia devuelta
 *      con una nota sobre si cargaron o no.
 */
import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (categoryName) => {
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

	useEffect(() => {
		getGifs(categoryName)
			.then((imgs) => {
				setState({
					data: imgs,
					loading: false,
				});
			})
			.catch((error) => {
				console.error(error);
			});
	}, [categoryName]);

	return state;
};
