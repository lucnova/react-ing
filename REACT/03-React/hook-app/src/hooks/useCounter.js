import { useState } from 'react';

// CUSTOM HOOK Contador
// Este custom hook extrae la lÃ³gica del contador anterior.

export const useCounter = (initialState = 0) => {
	const [count, setCount] = useState(initialState);

	const increment = (step) => {
		setCount(count + step);
	};

	const decrement = (step) => {
		setCount(count - step);
	};

	const reset = () => {
		setCount(initialState);
	};

	// El Custom Hook retorna:
	return {
		count,
		increment,
		decrement,
		reset,
	};
};
