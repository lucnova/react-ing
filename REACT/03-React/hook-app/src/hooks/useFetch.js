import { useEffect, useState } from 'react';

const loadingState = {
	data: null,
	loading: true,
	error: null,
};

export const useFetch = (url = '') => {
	const [state, setState] = useState(loadingState);

	const sendRequest = () => {
		setState(loadingState);

		fetch(url)
			.then((response) => response.json())
			.then((JSONData) => {
				setState({
					data: JSONData,
					error: null,
					loading: false,
				});
			})
			.catch((e) => {
				console.error(e);
			});

		return () => {
			//cleanup
		};
	};

	useEffect(() => {
		sendRequest();
	}, [url]);

	return [state, sendRequest];
};
