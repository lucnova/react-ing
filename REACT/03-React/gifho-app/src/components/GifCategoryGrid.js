import React /*  useState, useEffect */ from 'react'; // useEffect -> Ejecutar c칩digo de manera condicional
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { GifCategoryGridItem } from './GifCategoryGridItem';
// import { getGifs } from '../helpers/getGifs';

const GifCategoryGrid = ({ categoryName }) => {
	// const [imgs, setImgs] = useState([]);
	const {loading, data} = useFetchGifs();

	console.log(loading);

	//  - USE EFECT -
	//      Evita renderizar todo el c칩digo de nuevo para cuando hay cambios peque침os
	//      -> (callback, lista de dependencias)
	//  Si se manda vacia la lista de dependencias entonces solo se ejecuta 1 vez
	// useEffect(() => {
	// 	getGifs(categoryName)
	// 		.then(setImgs)
	// 		.catch((error) => {
	// 			console.error(error);
	// 		});
	// }, [categoryName]); // El arreglo contiene variables de las cuales useEffect se ejecutar치 denuevo

	return (
		<>
			<h3>{categoryName}</h3>
			{/* <div className="card-grid">
				{imgs.map((img) => (
					<GifCategoryGridItem key={img.id} {...img} />
				))}
			</div> */}
			{loading ? 'Loading...' : 'DONE'}
		</>
	);
};

GifCategoryGrid.propTypes = {
	categoryName: PropTypes.string.isRequired,
};

export default GifCategoryGrid;
