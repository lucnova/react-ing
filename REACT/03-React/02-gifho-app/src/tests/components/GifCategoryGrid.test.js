import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import GifCategoryGrid from '../../components/GifCategoryGrid';

/* Importo el CUSTOM HOOK */
import { useFetchGifs } from '../../hooks/useFetchGifs';
/* Realizo un mock con jest para poder 'inicializarlo' */
jest.mock('../../hooks/useFetchGifs');

describe('testing on GifCategoryGrid', () => {
	const sentCategoryName = 'DDLC';

	test('should match the component snapshot', () => {
		useFetchGifs.mockReturnValue({
			data: [],
			loading: true,
		});

		const wrapper = shallow(<GifCategoryGrid categoryName={sentCategoryName} />);

		expect(wrapper).toMatchSnapshot();
	});

	test('should display GridItems when data is gathered by useFetchGifs', () => {
		const imgs = [
			{
				id: '1241',
				title: 'tenor.com',
				url: 'https://media.tenor.com/images/bcb63018f904222a7f17e0532a517269/tenor.gif',
			},
			{
				id: '5346',
				title: 'tenor.com',
				url: 'https://media.tenor.com/images/bcb63018f904222a7f17e0532a517269/tenor.gif',
			},
		];

		useFetchGifs.mockReturnValue({
			data: imgs,
			loading: false,
		});

		const wrapper = shallow(<GifCategoryGrid categoryName={sentCategoryName} />);

		// Debe ser igual a snapshot
		expect(wrapper).toMatchSnapshot();
		// NO debe de mostrar el elemento loading
		expect(wrapper.find('p.animate__animated').exists()).toBe(false);
		// Debe mostrar la misma cantidad de componentes que del arreglo de data
		expect(wrapper.find('GifCategoryGridItem').length).toBe(imgs.length);
	});
});
