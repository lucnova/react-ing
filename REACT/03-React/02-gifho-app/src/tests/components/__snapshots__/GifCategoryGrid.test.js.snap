// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing on GifCategoryGrid should display GridItems when data is gathered by useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInUp"
  >
    DDLC
  </h3>
  <div
    className="card-grid"
  >
    <GifCategoryGridItem
      id="1241"
      key="1241"
      title="tenor.com"
      url="https://media.tenor.com/images/bcb63018f904222a7f17e0532a517269/tenor.gif"
    />
    <GifCategoryGridItem
      id="5346"
      key="5346"
      title="tenor.com"
      url="https://media.tenor.com/images/bcb63018f904222a7f17e0532a517269/tenor.gif"
    />
  </div>
</Fragment>
`;

exports[`testing on GifCategoryGrid should match the component snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInUp"
  >
    DDLC
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
